File: app/src/main/java/wannabit/io/cosmostaion/fragment/NumberKeyBoardFragment.java
package wannabit.io.cosmostaion.fragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.ImageButton;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Random;

import wannabit.io.cosmostaion.R;

public class NumberKeyBoardFragment extends KeyboardFragment implements View.OnClickListener {

    private View rootView;
    private final Button[] numberButtons = new Button[10];
    private ArrayList<String> numberArray = new ArrayList<>();

    public static NumberKeyBoardFragment newInstance() {
        return new NumberKeyBoardFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_keyboard_number, container, false);
        numberArray = new ArrayList<>(Arrays.asList(getResources().getStringArray(R.array.password_number)));
        Collections.shuffle(numberArray, new Random(System.nanoTime()));
        final String packageName = rootView.getContext().getPackageName();
        for (int i = 0; i < numberButtons.length; i++) {
            numberButtons[i] = rootView.findViewById(getResources().getIdentifier("password_number" + i, "id", packageName));
            numberButtons[i].setText(numberArray.get(i));
            numberButtons[i].setOnClickListener(this);
        }
        rootView.findViewById(R.id.password_back).setOnClickListener(this);
        return rootView;
    }

    @Override
    public void shuffleKeyboard() {
        final Animation fadeInAnimation = AnimationUtils.loadAnimation(requireContext(), R.anim.fade_in);
        fadeInAnimation.reset();

        Animation fadeOutAnimation = AnimationUtils.loadAnimation(requireContext(), R.anim.fade_out);
        fadeOutAnimation.reset();
        fadeOutAnimation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {
            }

            @Override
            public void onAnimationEnd(Animation animation) {
                Collections.shuffle(numberArray, new Random(System.nanoTime()));
                for (int i = 0; i < numberButtons.length; i++) {
                    numberButtons[i].setText(numberArray.get(i));
                }
                rootView.startAnimation(fadeInAnimation);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {
            }
        });
        rootView.startAnimation(fadeOutAnimation);

    }

    @Override
    public void onClick(View view) {
        if (view instanceof Button) {
            if (keyboardListener != null) {
                keyboardListener.userInsertKey(((Button) view).getText().toString().trim().toCharArray()[0]);
            }

        } else if (view instanceof ImageButton) {
            if (keyboardListener != null) {
                keyboardListener.userDeleteKey();
            }
        }
    }
}

File: app/src/main/java/wannabit/io/cosmostaion/ui/password/NumberKeyBoardFragment.kt
package wannabit.io.cosmostaion.ui.password

import android.content.Context
import android.os.Build
import android.os.Bundle
import android.os.VibrationEffect
import android.os.Vibrator
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageButton
import wannabit.io.cosmostaion.R
import wannabit.io.cosmostaion.databinding.FragmentNumberKeyBoardBinding
import java.util.Random

class NumberKeyBoardFragment : KeyboardFragment(), View.OnClickListener {

    private var _binding: FragmentNumberKeyBoardBinding? = null
    private val binding get() = _binding!!

    private val numbers = mutableListOf<String>()
    private val numberBtns = arrayOfNulls<Button>(10)

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        _binding = FragmentNumberKeyBoardBinding.inflate(layoutInflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initView()
    }

    private fun initView() {
        binding.apply {
            val passwordList = resources.getStringArray(R.array.password_number)
            numbers.addAll(passwordList)
            numbers.shuffle(Random(System.nanoTime()))

            for (i in numberBtns.indices) {
                numberBtns[i] = requireView().findViewById(
                    resources.getIdentifier(
                        "password_number$i", "id", requireActivity().packageName
                    )
                )
                numberBtns[i]?.text = numbers[i]
                numberBtns[i]?.setOnClickListener(this@NumberKeyBoardFragment)
            }
            passwordBack.setOnClickListener(this@NumberKeyBoardFragment)
        }
    }

    override fun onClick(view: View?) {
        if (view is Button) {
            if (listener != null) {
                val input = view.text.toString().trim().toCharArray()[0]
                initVibrate()
                listener?.userInsertPassword(input)
            }

        } else if (view is ImageButton) {
            if (listener != null) {
                initVibrate()
                listener?.userDeletePassword()
            }
        }
    }

    private fun initVibrate() {
        val vibrator = requireActivity().getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(100, 50))
        } else {
            vibrator.vibrate(100)
        }
    }

    override fun onDestroyView() {
        _binding = null
        super.onDestroyView()
    }
}